sql/2020/11_Security/feature_adoption_by_country.sql

```diff
@@ -0,0 +1,48 @@
+#standardSQL
+# Security feature adoption grouped by sites frequently visited from different countries
+
+CREATE TEMP FUNCTION getNumSecurityHeaders(headers STRING) AS (
+  (
+    SELECT
+      COUNTIF(REGEXP_CONTAINS(headers, CONCAT('(?i)', headername)))
+    FROM 
+      UNNEST(['Content-Security-Policy', 'Cross-Origin-Embedder-Policy', 'Cross-Origin-Opener-Policy', 'Cross-Origin-Resource-Policy', 'Expect-CT', 'Feature-Policy', 'Permissions-Policy', 'Referrer-Policy', 'Report-To', 'Strict-Transport-Security', 'X-Content-Type-Options', 'X-Frame-Options', 'X-XSS-Protection']) AS headername
+  )
+);
+
+SELECT
+  client,
+  country,
+  COUNT(0) AS total_pages_for_country,
+  COUNTIF(STARTS_WITH(url, 'https')) AS freq_https,
+  SAFE_DIVIDE(COUNTIF(STARTS_WITH(url, 'https')), COUNT(0)) AS pct_https,
+  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, '(?i)X-Frame-Options')), COUNT(0)) AS pct_xfo,
+  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, '(?i)Strict-Transport-Security')), COUNT(0)) AS pct_hsts,
+  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, '(?i)X-Content-Type-Options')), COUNT(0)) AS pct_xcto,
+  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, '(?i)Expect-CT')), COUNT(0)) AS pct_expectct,
+  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, '(?i)Content-Security-Policy')), COUNT(0)) AS pct_csp,
+  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, '(?i)X-XSS-Protection')), COUNT(0)) AS pct_xss,
+  AVG(getNumSecurityHeaders(respOtherHeaders)) AS avg_security_headers,
+  APPROX_QUANTILES(getNumSecurityHeaders(respOtherHeaders), 1000)[OFFSET(500)] AS median_security_headers
```

Comment:
Not sure on point of this? The security header is either 0 or 1 so it's just returning that?

Reply:
it counts the number of security headers: so if a site has CSP + XFO, it will calculate as `2`